include/platform/mir/graphics/platform.h

```diff
@@ -71,11 +179,279 @@ class RenderingPlatform
     /**
      * Creates the buffer allocator subsystem.
      */
-    virtual UniqueModulePtr<GraphicBufferAllocator> create_buffer_allocator(
-        Display const& output) = 0;
+    virtual UniqueModulePtr<GraphicBufferAllocator> create_buffer_allocator(Display const& output) = 0;
+
+    /**
+     * Attempt to acquire a platform-specific interface from this RenderingPlatform
+     *
+     * Any given platform is not guaranteed to implement any specific interface,
+     * and the set of supported interfaces may depend on the runtime environment.
+     *
+     * Since this may result in a runtime probe the call may be costly, and the
+     * result should be saved rather than re-acquiring an interface each time.
+     *
+     * \tparam Interface
+     * \return  On success: an occupied std::shared_ptr<Interface>
+     *          On failure: std::shared_ptr<Interface>{nullptr}
+     */
+    template<typename Interface>
+    static auto acquire_interface(std::shared_ptr<RenderingPlatform> platform) -> std::shared_ptr<Interface>
+    {
+        static_assert(
+            std::is_convertible_v<Interface*, RendererInterfaceBase*>,
+            "Can only acquire a Renderer interface; Interface must implement RendererInterfaceBase");
+
+        if (auto const base_interface = platform->maybe_create_interface(typename Interface::Tag{}))
+        {
+            if (auto const requested_interface = std::dynamic_pointer_cast<Interface>(base_interface))
+            {
+                return requested_interface;
+            }
+            BOOST_THROW_EXCEPTION((
+                std::logic_error{
+                    "Implementation error! Platform returned object that does not support requested interface"}));
+        }
+        return nullptr;
+    }
+
+protected:
+    /**
+     * Acquire a specific hardware interface
+     *
+     * This should perform any runtime checks necessary to verify the requested interface is
+     * expected to work and return a pointer to an implementation of that interface.
+     *
+     * This function is guaranteed to be called with `this` managed by a `shared_ptr`; if
+     * the returned value needs to share ownership with `this`, calls to std::shared_from_this
+     * can be expected to work.
+     *
+     * \param type_tag  [in]    An instance of the Tag type for the requested interface.
+     *                          Implementations are expected to dynamic_cast<> this to
+     *                          discover the specific interface being requested.
+     * \return      A pointer to an implementation of the RenderInterfaceBase-derived
+     *              interface that corresponds to the most-derived type of tag_type.
+     */
+    virtual auto maybe_create_interface(
+        RendererInterfaceBase::Tag const& type_tag) -> std::shared_ptr<RendererInterfaceBase> = 0;
```

Comment:
Overly generic: This is only ever called with `GLRenderingProvider::Tag{}` and the return type is always downcast to `shared_ptr<GLRenderingProvider>`